torch/nn/utils/_expanded_weights/expanded_weights_utils.py

```diff
@@ -0,0 +1,114 @@
+        num_true_outs: The number of outputs seen by the user since some functions
+          return auxillary data that is only used in the backward pass
+    '''
+    unexpanded_args = _check_and_unexpand_args(func, expanded_args)
+    output = func(*unexpanded_args)
+    output, aux_outputs = _check_and_detach_output(output, num_true_outs)
+    return (output, expanded_args, aux_outputs)
+
+def _check_and_unexpand_args(func, expanded_args):
+    # input must be the first argument passed
+    input = expanded_args[0]
+    if isinstance(input, ExpandedWeight):
+        raise RuntimeError("Expanded Weights do not support inputs that are also ExpandedWeights. "
+                           f"Input must be a Tensor, got {type(input).__name__} in function {func.__name__}")
+    if not isinstance(input, torch.Tensor):
+        raise RuntimeError("Expanded Weights requires a Tensor as the first input to get the batch dimension, "
+                           f"got {type(input).__name__} in function {func.__name__}")
+    if len(input.shape) == 0:
+        raise RuntimeError(f"Expanded Weights requires a batch dimension but got an input of size 0 in function {func.__name__}")
+    if input.shape[0] == 0:
+        raise RuntimeError("0 is not a valid batch size for Expanded Weights but got input tensor of "
+                           f"{input} in function {func.__name__}")
+    batch_size = input.shape[0]
+    for arg in expanded_args:
+        if isinstance(arg, ExpandedWeight) and arg.batch_size != batch_size:
+            raise RuntimeError("Expected ExpandedWeights to have batch size matching input but got "
+                               f"input batch size of {batch_size} with ExpandedWeight of batch size {arg.batch_size}")
+
+    unexpanded_args = tuple(arg.orig_weight if isinstance(arg, ExpandedWeight) else arg for arg in expanded_args)
+    return unexpanded_args
+
+def _check_and_detach_output(output, num_true_outs):
+    aux_outputs = None
+    # separates differentiable outputs from outputs only needed for the backwards computation
+    if isinstance(output, tuple):
+        if len(output) < num_true_outs:
+            raise RuntimeError(f"Got fewer outputs ({len(output)}) than expected ({num_true_outs}). "
+                               "Issues in ExpandedWeights' autograd.Function")
+        aux_outputs = output[num_true_outs:]
+        output = output[:num_true_outs]
+    elif num_true_outs != 1:
+        raise RuntimeError(f"Got single output but expected at least {num_true_outs} outputs. "
+                           "Issues in ExpandedWeights' autograd.Function")
+    return (output, aux_outputs)
+
+def set_grad_sample_if_exists(maybe_expanded_weight, per_sample_grad_fn):
+    unpacked = unpack_expanded_weight_or_tensor(maybe_expanded_weight)
+    if isinstance(maybe_expanded_weight, ExpandedWeight):
+        if hasattr(unpacked, "grad_sample"):
+            if isinstance(unpacked.grad_sample, list):
```

Comment:
This will put the grad_samples together in a list if there's more than one. This follows what Opacus does and has some nice flexibilit. Typically you will want the stacked gradients (this is now done in a single stack versus having to do multiple stacks if we returned a tensor, not a list), but in the case of RNNs, users will want the per sample gradients to be accumulated for RNN cell (Opacus has a special layer to do this). We would still be able to support accumulation behavior for RNNs, the summing would just be taken care of by Opacus instead of autograd

Results still showed significant speedup for ExpandedWeights over Opacus, even though we're saving more memory

Reply:
Also added a couple small tests that repeats the tests checking the per sample gradients, but is called twice with the same inputs. The test checks that the behavior matched what would happen if we had a single batch with the inputs repeated twice